lib/behavior.js

```diff
@@ -0,0 +1,170 @@
+'use strict';
+
+/**
+ * Built-in behaviors.
+ * These provide hooks into modifying Mocha's behavior for different use cases.
+ */
+
+var utils = require('./utils');
+
+/**
+ * Default Mocha behavior
+ */
+exports.Default = {
+  Suite: {
+    /**
+     * Runs the suite; called by `Suite.create()`.  Calls its callback fn
+     * with the Suite as its context.
+     * @this {Suite}
+     */
+    run: function() {
+      // FUTURE: async suites
+      this.fn.apply(this);
+    }
+  },
+  Runnable: {
+    /**
+     * Determines whether or not we should provide a nodeback parameter and
+     * expect it to be called
+     * @this {Runnable}
+     * @returns {boolean} `true` if we should run the test fn with a callback
+     */
+    shouldRunWithCallback: function() {
+      return Boolean(this.fn && this.fn.length);
+    },
+    /**
+     * Runs the Runnable synchronously, or potentially returning a Promise
+     * @param {Function} done - Callback
+     */
+    run: function(done) {
+      var result = this.fn.call(this.ctx);
+      if (result && typeof result.then === 'function') {
```

Comment:
I know this code is just moved, but if we wrap the result in `Promise.resolve(result).then(...)` instead of doing this check, we'll always have the returned value treated as a promise, without the need for the extra hassle.

Reply:
This would require adding a Promise polyfill to the production bundle.

Reply:
(I'm not necessarily offering an opinion on that; I'm just telling you why Mocha historically does it this way)

Reply:
I'm actually really annoyed about IE11/ES5, so maybe we can just write Mocha in ES6 going forward and pipe the distfile thru Babel...